satpy/readers/yaml_reader.py

```diff
@@ -1374,22 +1354,141 @@ def _get_empty_segment_with_height(empty_segment, new_height, dim):
     return empty_segment
 
 
-def _get_FCI_L1c_FDHSI_chunk_height(chunk_width, chunk_n):
-    """Get the height in pixels of a FCI L1c FDHSI chunk given the chunk width and number (starting from 1)."""
-    if chunk_width == 11136:
-        # 1km resolution case
-        if chunk_n in [3, 5, 8, 10, 13, 15, 18, 20, 23, 25, 28, 30, 33, 35, 38, 40]:
-            chunk_height = 279
-        else:
-            chunk_height = 278
-    elif chunk_width == 5568:
-        # 2km resolution case
-        if chunk_n in [5, 10, 15, 20, 25, 30, 35, 40]:
-            chunk_height = 140
-        else:
-            chunk_height = 139
-    else:
-        raise ValueError("FCI L1c FDHSI chunk width {} not recognized. Must be either 5568 or 11136.".format(
-            chunk_width))
+class GEOVariableSegmentYAMLReader(GEOSegmentYAMLReader):
+    """GEOVariableSegmentYAMLReader for handling chunked/segmented GEO products with segments of variable height.
 
-    return chunk_height
+    This YAMLReader ovverrides parts of the GEOSegmentYAMLReader to account for formats where the segments can
+    have variable heights. It computes the sizes of the padded segments using the information available in the
+    file(handlers), so that gaps of any size can be filled as needed.
+
+    This implementation was motivated by the FCI L1c format, where the segments (called chunks in the FCI world)
+    can have variable heights. It is however generic, so that any future reader can use it. The requirement
+    for the reader is to have a method called `get_segment_position_info`, returning a dictionary containing
+    the positioning info for each chunk (see example in
+    :func:`satpy.readers.fci_l1c_nc.FCIL1cNCFileHandler.get_segment_position_info`).
+
+    For more information on please see the documentation of :func:`satpy.readers.yaml_reader.GEOSegmentYAMLReader`.
+    """
+
+    def create_filehandlers(self, filenames, fh_kwargs=None):
+        """Create file handler objects and collect the location information."""
+        created_fhs = super().create_filehandlers(filenames, fh_kwargs=fh_kwargs)
+        self._extract_segment_location_dicts(created_fhs)
+        return created_fhs
+
+    def _extract_segment_location_dicts(self, created_fhs):
+        self.segment_infos = dict()
+        for filetype, filetype_fhs in created_fhs.items():
+            # initialise the segment info for this filetype
+            exp_segment_nr = filetype_fhs[0].filetype_info['expected_segments']
+            width_to_grid_type = _get_width_to_grid_type(filetype_fhs[0].get_segment_position_info())
+            self.segment_infos.update({filetype: {'available_segment_infos': [],
+                                                  'expected_segments': exp_segment_nr,
+                                                  'width_to_grid_type': width_to_grid_type}})
```

Comment:
This could be a separate method, for example `self._extract_initial_segment_position_info(filetype, filetype_fhs)`.